pkg/filter/kafka/kafka.go

```diff
@@ -125,22 +141,55 @@ func (k *Kafka) Status() interface{} {
 
 // HandleMQTT handle MQTT context
 func (k *Kafka) HandleMQTT(ctx context.MQTTContext) *context.MQTTResult {
-	if ctx.PacketType() != context.MQTTPublish {
-		return &context.MQTTResult{}
+	var topic string
+	var headers map[string]string
+	var payload []byte
+	var ok bool
+
+	// set data from kv map
+	if k.topic != "" {
+		topic, ok = ctx.GetKV(k.topic).(string)
+		if !ok {
+			return &context.MQTTResult{ErrString: resultGetDataFailed}
+		}
+	}
+	if k.headers != "" {
+		headers, ok = ctx.GetKV(k.headers).(map[string]string)
+		if !ok {
+			return &context.MQTTResult{ErrString: resultGetDataFailed}
+		}
+	}
+	if k.payload != "" {
+		payload, ok = ctx.GetKV(k.payload).([]byte)
+		if !ok {
+			return &context.MQTTResult{ErrString: resultGetDataFailed}
+		}
 	}
 
-	p := ctx.PublishPacket()
-	logger.Debugf("produce msg with topic %s", p.Topic())
+	// set data from PublishPacket if data is missing
```

Comment:
Payload in the Context is higher priority then PublishPacket

Reply:
updated!